A71-xds-fallback.md

```diff
@@ -0,0 +1,226 @@
+by channel target.
+1. Change xDS bootstrap code to return the full list of servers for each
+xDS authority from the bootstrap config.
+1. Change XdsClient to support fallback within each xDS authority.
+
+### Change global XdsClient from a single instance to per-target instances
+
+Current gRPC implementation is using a single XdsClient for all data plane
+targets. For fallback, if any one channel triggers fallback, then all channels
+regardless of authority would switch to using the data from the fallback xDS
+control plane which is not desirable.
+
+gRPC servers using the xDS configuration will share the same XdsClient instance
+keyed with a dedicated well-known key value.
+
+Changes XdsClient to per-data plane target would enable gRPC to switch to
+fallback configuration only for channels that need xDS resources that have not
+been downloaded yet.
+
+This change would change the channels to get XdsClient for their data plane
+target. It will also require changes to the CSDS service to aggregate configs
+from multiple XdsClients. The CSDS RPC service will be changed to get data
+from all XdsClient instances. A new string field named `channel_target` will be
+added to the CSDS `ClientConfig` message to indicate the channel target
+the data is associated with.
+
+Adding this field will not require updating the CSDS client as its value will
+be the same for most resources in most cases.
+
+### Change xDS Bootstrap to handle multiple xDS configuration servers
+
+Current gRPC xDS implementations only use the first xDS configuration server
+listed in the bootstrap JSON document. Fallback implementation requires changes
+to use all servers listed in order from highest to lowest priority.
+
+### Change XdsClient to support fallback within each xDS authority
+
+The fallback process is initiated if both of the following conditions hold:
+
+* There is a connectivity failure on the ADS stream, as described in [A57] this
+means either:
+    - Channel reports TRANSIENT_FAILURE.
+    - The ADS stream was closed before the first server response had been
+    received.
+* At least one watcher exists for a resource that is not cached. Cached
+resources include:
+    - Resources that have been successfully received and can be used.
+    - Resources that are considered non-existent according
+    to [xDS Protocol Specification][resource-does-not-exist].
+
+This may happen during a new channel setup or mid-config update, when
+the control plane notifies of the config change and becomes unavailable before
+the new config resources are downloaded.
+
+XdsClients will need to be changed to support multiple ADS connections for each
+authority. Once the fallback process begins, an impacted XdsClient will establish
+a connection to the next xDS control plane listed in the bootstrap JSON. Then
+XdsClient will subscribe to all watched resources on that server and will
+update the cache based on the received responses.
```

Comment:
Can we add more details here with example configurations?

It is not clear to me what happens in the following scenario
- a grpc channel has received its full configuration and is serving traffic
- a config update adds a new cluster in the route table
- at the same time the xDS channel moves to TRANSIENT_FAILURE
- Do we switch to the fallback server in the background and wait to receive the full configuration before using it, or do we switch immediately and fail RPCs until we receive the full configuration

The same question applies when a grpc channel has received full configuration from a secondary xds server, and the primary becomes available. When do we switch to it?

Reply:
> a config update adds a new cluster in the route table
> at the same time the xDS channel moves to TRANSIENT_FAILURE

Depends on whether the channel is trying to obtain updated resource. E.g. there's an inherent race condition whether or not the channel "knows" there was a config change.

> The same question applies when a grpc channel has received full configuration from a secondary xds server, and > the primary becomes available. When do we switch to it?

Switch happens as soon as a connection to primary server is reestablished.

Reply:
> It is not clear to me what happens in the following scenario
> 
> * a grpc channel has received its full configuration and is serving traffic
> * a config update adds a new cluster in the route table
> * at the same time the xDS channel moves to TRANSIENT_FAILURE

(This is basically the mid-update case referenced in the gRFC, but as per my comment elsewhere, we need to clarify the description.)

> * Do we switch to the fallback server in the background and wait to receive the full configuration before using it, or do we switch immediately and fail RPCs until we receive the full configuration

Neither.  What will happen is that the xds resolver will immediately get the new RDS resource, so it will add the CDS LB policy for the new cluster, and it will return a ConfigSelector that uses the new RouteConfig, so RPCs will immediately start being routed to that cluster.  However, the new CDS LB policy will not immediately get any notification on its CDS watch, so it will stay in CONNECTING state until the XdsClient switches over to the fallback server and obtains the CDS resource from there.  The result is that we won't fail any RPCs, but we will cause some latency.

The ideal behavior is that the channel would not switch over to the new RouteConfig until it obtains the necessary CDS resource from the fallback server, so that we don't cause that unnecessary latency.  However, we can't do that until we implement gRFC A74 (https://github.com/grpc/proposal/pull/404), which we decided to split out from this gRFC, since it's really a somewhat separate issue.

> The same question applies when a grpc channel has received full configuration from a secondary xds server, and the primary becomes available. When do we switch to it?

As soon as we get the first resource from the primary server, we immediately close the connection to the fallback server, and we update the watchers from the responses we get from the primary server.

The idea here is that the fallback logic is entirely within the XdsClient and is not visible to the XdsClient's watchers in the gRPC channel.  The logic in the gRPC channel will just continue to make decisions the way it always has.